src/pages/experiments/[experimentId]/data-processing/index.jsx

```diff
@@ -199,126 +255,216 @@ const DataProcessingPage = ({ experimentId, experimentData, route }) => {
   }, [stepIdx]);
 
   useEffect(() => {
-    if (!completingStepError) {
+    if (completingStepError && stepIdx > completedSteps.size) {
+      setStepIdx(completedSteps.size);
       return;
     }
 
-    if (stepIdx > completedSteps.size) {
-      setStepIdx(completedSteps.size);
-    }
   }, [completingStepError, stepIdx]);
 
+  const manualStepIdxChange = (newStepIdx) => {
+    if (changesOutstanding) {
+      upcomingStepIdxRef.current = newStepIdx;
+
+      setShowChangesWillBeLost(true);
+    } else {
+      setStepIdx(newStepIdx);
+    }
+  }
+
+  let nextDisabledByPipeline = false;
+
+  if (steps[stepIdx + 1]) {
+    const { key: nextKey } = steps[stepIdx + 1];
+
+    // We disable the next 
+    nextDisabledByPipeline = pipelineBlockingSteps && !stepIsCompletedInPipeline(nextKey) && !completedSteps.has(nextKey);
+  }
+
+  const completeStepAt = (stepIndex) => {
+    const newDoneStepKey = steps[stepIndex].key;
+
+    const newStepsDone = new Set([...completedSteps, newDoneStepKey]);
+
+    dispatch(updateCompletedSteps(experimentId, newStepsDone, steps.length));
+  }
+
+  const setCompletedStepsFrom = (calledFromStepKey) => {
+    let newStepsDone = [];
+
+    steps.forEach((step) => {
+      newStepsDone.push(step.key);
+
+      if (step.key === calledFromStepKey) {
+        dispatch(updateCompletedSteps(experimentId, newStepsDone, steps.length))
+        return;
+      }
+    });
+  }
```

Comment:
Are you sure this is a good name for the function? I don't think it does what it says, i.e. setting all steps to be complete /up to/ the key you give it. Or am I mistaken?

Reply:
I decided to leave it in a simple form.
It receives an array of keys and all it does is set these keys as `stepsDone`.

Reply:
Maybe a better name could be `setCompletedSteps`?

Reply:
Sure, that sounds reasonable.